File: packages/sdk-py/examples/webhook_subscription.py
"""Register a webhook subscription and print the signing secret.

The signing secret is returned by the API exactly once — persist it
immediately (e.g. to a secret manager). Running this example twice
creates two subscriptions.

Usage::

    python examples/webhook_subscription.py <project_slug> <name> <url> [events…]

Example::

    python examples/webhook_subscription.py demo "Slack #deploys" \\
        https://hooks.slack.com/services/T/B/X \\
        deploy.succeeded deploy.failed rollback.succeeded

Environment variables:
    ENCLII_TOKEN     Required. API token.
    ENCLII_API_URL   Optional. Defaults to https://api.enclii.dev.
"""

from __future__ import annotations

import asyncio
import os
import sys

from enclii_sdk import AsyncEncliiClient


async def main(project_slug: str, name: str, url: str, events: list[str]) -> int:
    if not events:
        # Sensible default for most callers.
        events = ["deploy.succeeded", "deploy.failed", "rollback.succeeded"]

    async with AsyncEncliiClient(
        base_url=os.environ.get("ENCLII_API_URL", "https://api.enclii.dev"),
        token=os.environ["ENCLII_TOKEN"],
    ) as enclii:
        resp = await enclii.webhooks.create(
            project_slug, name=name, url=url, events=events
        )
        sub = resp.subscription
        print(f"Created subscription {sub.id}")
        print(f"  name:   {sub.name}")
        print(f"  url:    {sub.url}")
        print(f"  events: {', '.join(e.value for e in sub.event_types)}")
        print()
        print("SIGNING SECRET (shown once — save this now):")
        print(f"  {resp.signing_secret}")
        if resp.note:
            print()
            print(resp.note)
    return 0


if __name__ == "__main__":
    if len(sys.argv) < 4:
        print(__doc__)
        sys.exit(2)
    sys.exit(
        asyncio.run(main(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4:]))
    )

File: packages/sdk-ts/examples/webhook-subscription.ts
/**
 * Create an outbound lifecycle webhook subscription and verify a signed
 * delivery. Illustrates the "signing secret returned exactly once" contract.
 *
 * Usage:
 *   ENCLII_TOKEN=... ENCLII_PROJECT=my-project \
 *     ENCLII_WEBHOOK_URL=https://hooks.example.com/enclii \
 *     tsx examples/webhook-subscription.ts
 */

import {
  EncliiClient,
  verifyWebhookSignature,
} from '@madfam/enclii-sdk';

async function main() {
  const client = new EncliiClient({
    baseUrl: process.env.ENCLII_BASE_URL ?? 'https://api.enclii.dev/v1',
    token: process.env.ENCLII_TOKEN,
  });

  const project = req('ENCLII_PROJECT');
  const url = req('ENCLII_WEBHOOK_URL');

  // 1. Create the subscription. This returns the raw signing secret exactly once.
  const { subscription, signing_secret } = await client.webhooks.create(
    project,
    {
      name: 'example-subscription',
      url,
      event_types: ['deploy.succeeded', 'deploy.failed', 'rollback.succeeded'],
    },
  );
  console.log(`Created subscription ${subscription.id}`);
  console.log(`URL:          ${subscription.url}`);
  console.log(`Events:       ${subscription.event_types.join(', ')}`);
  console.log();
  console.log('=== SIGNING SECRET (save now — never shown again) ===');
  console.log(signing_secret);
  console.log();

  // 2. Trigger a synthetic test.ping to verify the endpoint handles our
  //    signed deliveries correctly.
  console.log('Firing test.ping...');
  const delivery = await client.webhooks.test(subscription.id);
  console.log(`Test delivery ${delivery.id} enqueued.`);

  // 3. In a real receiver, you'd verify signatures like this. The raw body is
  //    the exact bytes you received — do NOT re-serialize a parsed JSON object,
  //    whitespace differences break the HMAC.
  const exampleRawBody = '{"example":true}';
  const exampleTimestamp = Math.floor(Date.now() / 1000);
  const exampleSig = await computeSignatureFixture(
    signing_secret,
    exampleTimestamp,
    exampleRawBody,
  );
  await verifyWebhookSignature(
    exampleRawBody,
    `t=${exampleTimestamp},v1=${exampleSig}`,
    signing_secret,
  );
  console.log('Signature verification round-trip works.');
}

async function computeSignatureFixture(
  secret: string,
  ts: number,
  body: string,
): Promise<string> {
  const key = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode(secret),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign'],
  );
  const sig = await crypto.subtle.sign(
    'HMAC',
    key,
    new TextEncoder().encode(`${ts}.${body}`),
  );
  return Array.from(new Uint8Array(sig))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

function req(name: string): string {
  const v = process.env[name];
  if (!v) {
    console.error(`missing ${name}`);
    process.exit(64);
  }
  return v;
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
